core/src/font.rs

```diff
@@ -60,36 +60,94 @@ impl EvalParameters {
+        /// Used by `DefineEditText` tags.
+        code_point_to_glyph: fnv::FnvHashMap<u16, usize>,
+
+        /// Kerning infomration.
+        /// Maps from a pair of unicode code points to horizontal offset value.
+        kerning_pairs: fnv::FnvHashMap<(u16, u16), Twips>,
+    },
+    Empty,
+}
+
+impl GlyphSource {
+    pub fn get_by_index(&self, index: usize) -> Option<&Glyph> {
+        match self {
+            GlyphSource::Memory { glyphs, .. } => glyphs.get(index),
+            GlyphSource::Empty => None,
+        }
+    }
+
+    pub fn get_by_code_point(&self, code_point: char) -> Option<&Glyph> {
+        match self {
+            GlyphSource::Memory {
+                glyphs,
+                code_point_to_glyph,
+                ..
+            } => {
+                // TODO: Properly handle UTF-16/out-of-bounds code points.
+                let code_point = code_point as u16;
+                if let Some(index) = code_point_to_glyph.get(&code_point) {
+                    glyphs.get(*index)
+                } else {
+                    None
+                }
+            }
+            GlyphSource::Empty => None,
+        }
+    }
+
+    pub fn has_kerning_info(&self) -> bool {
+        match self {
+            GlyphSource::Memory { kerning_pairs, .. } => !kerning_pairs.is_empty(),
+            GlyphSource::Empty => false,
+        }
+    }
+
+    pub fn get_kerning_offset(&self, left: char, right: char) -> Twips {
+        match self {
+            GlyphSource::Memory { kerning_pairs, .. } => {
+                // TODO: Properly handle UTF-16/out-of-bounds code points.
+                let left_code_point = left as u16;
+                let right_code_point = right as u16;
+                kerning_pairs
+                    .get(&(left_code_point, right_code_point))
+                    .cloned()
+                    .unwrap_or_default()
+            }
+            GlyphSource::Empty => Twips::ZERO,
+        }
+    }
+}
+
 #[derive(Debug, Clone, Collect, Copy)]
 #[collect(no_drop)]
 pub struct Font<'gc>(Gc<'gc, FontData>);
 
-#[derive(Debug, Clone, Collect)]
+#[derive(Debug, Collect)]
 #[collect(require_static)]
 struct FontData {
-    /// The list of glyphs defined in the font.
-    /// Used directly by `DefineText` tags.
-    glyphs: Vec<Glyph>,
-
-    /// A map from a Unicode code point to glyph in the `glyphs` array.
-    /// Used by `DefineEditText` tags.
-    code_point_to_glyph: fnv::FnvHashMap<u16, usize>,
+    glyphs: GlyphSource,
 
     /// The scaling applied to the font height to render at the proper size.
     /// This depends on the DefineFont tag version.
     scale: f32,
 
-    /// Kerning infomration.
-    /// Maps from a pair of unicode code points to horizontal offset value.
-    kerning_pairs: fnv::FnvHashMap<(u16, u16), Twips>,
-
     /// The distance from the top of each glyph to the baseline of the font, in
     /// EM-square coordinates.
-    ascent: u16,
+    ascent: i16,
```

Comment:
Out of interest, why is this change necessary, can the ascent be negative?

Reply:
For TTFs yes, apparently